### ISSUE TITLE
security/audit: upstream merge processes

### ISSUE THREAD

**Issue by whilei** [ethoxy/multi-geth#35]
Currently upstream merges (incoming ETH) are managed by @sorpaas or myself, often independently.

As exemplified in #28, pulling in large and complex upstream changes can have unforeseen effects on behavior.

We need a better process for auditing these merges so that incoming code can be reviewed, knowledge about merge conflicts and their resolutions shared, and improved testing/auditing before upstream changes are allowed to hit master.

This might include requiring 1+ review approvals before allow to merge to master, some kind of CONFLICTLOG for documenting merge conflicts and their resolutions, code owners (as in #34)...

**Comment by soc1c** [comment]
I would propose a staging release process such as Parity Ethereum does. Having a nightly branch where we can experiment with new features and weekly downstream merges from go-ethereum and a "stable" branch and/or a "beta"/staging branch releases every 2-3 months. Breakage will be noticed faster and does not affect stable releases directly. And there is plenty of time to react and fix it.

Happy to assist.

**Comment by whilei** [comment]
Yea, thanks @soc1c . Sorry to be late responding to this. 

There's a lot of options we can explore if we start sticking our toes in to not-just-`master` schemes; nightlies, staging, release branches, tag branches, ... 

I'm personally not so keen on tiered branch strategies, with the reasons revolving around additional ambiguity and complexity../

Let's assume we use a tiered `nightly/beta/stable` branching strategy. Here's some issues:
- Do we roll nightly in to beta every... night? Every week? Or does it have to pass an additional set of tests that typical PRs to master wouldn't? If we do roll incrementally (time frame) the meaning of "beta" is then just really more "weekly and probably with some use in the wild by early-adopters who don't get much information about what they're adopting except that it happens every week."
- How, why, and when then to roll 'beta' in to 'stable' -- after a few months of "use" w/o "too many big" bug reports? Guaranteeing stability on absence of evidence seems backwards, and w/o clear metrics of use (at this point) we'd have a hard time judging even the scope of the absent non-negative feedback.

I guess the general point is that the edges between `nightly/beta/stable` seem to me to fall quickly into arbritrary and subjective implied meanings, and there's risk involved in labeling stuff w/o being able to clearly define or explain the meaning.

On the other hand, one of the consequences of this pattern would be that "stable" releases/branches would fall 2-3 months behind upstream or what otherwise might hit master. So we'd have to assume that although we wouldn't have a precise measure of how much, we could expect to have _more_ 'in wild' testing before a version was declared stable. 

But to that point, isn't that something that can be accomplished with the existing master/released-tag pattern? (Either with 'beta'/'stable' release drafts and/or tags).

I know this has been long and rambling; I apologize.

TL;DR: I think maybe we can accomplish the same effects of your proposed `feat/nightly/stable/release` branching scheme w/ the existing `feat/master + tag/release` scheme, which appeals to me because I think using git for branching schemes beyond as-simple-as-possible gets awkward and fragile quickly.

**Comment by meowsbits** [comment]
Just updating with some later thoughts on this. 

Having now gotten some experience with these merges (eg. https://github.com/multi-geth/multi-geth/pull/70, and sorpaas moving thru the infamous 1.8.24-27 series), I think we should try to emulate upstream go-ethereum's branch/release (erm, release-branch) pattern. Since one of our current priorities is to be able to dance neatly in near-lockstep with upstream, it seems following those patterns will simplify things.

They use a `master` branch and `release/X.Y` branches. Typically the master branch is rather diverged and ahead of the latest release branch.

This raises the question of releases, and how we want to manage releases in relation to their releases. It's probably safe to say that when they post a release, we should post a congruent release. But we will (and do already) have changes of our own that we'll want to release probably independent of their schedule. 
- Should those use their own branches?
- How should those tags be namespaced?  

How and if to "robot-ize" this?... I'm not sure.
